### ISSUE TITLE
Unexpected pinned_memory allocation

### ISSUE THREAD

**Issue by jemiryguo** [cupy/cupy#8090]
### Description

When using `cupy.asarray` to transfer a numpy array, which is allocated by `cupyx.empty_pinned`, into GPU, no more pinned_memory should be allocated. However, this is not the case. And this extra pinned_memory can not be free by `free_all_blocks()`.

See the code bellow for details.

### To Reproduce

```py
import cupy, cupyx, psutil, os

pool = cupy.get_default_pinned_memory_pool()
print("Pinned-memory Pool Free blocks:", pool.n_free_blocks(), "Total used CPU memory:", psutil.Process(os.getpid()).memory_info().rss/1024/1024/1024, "GB")
# output: 0, 0G

# allocate 8G pinned_memory
a = cupyx.empty_pinned((1024,1024,1024), dtype='f8')
print("a = cupyx.empty_pinned((1024,1024,1024), dtype='f8')")
print("Pinned-memory Pool Free blocks:", pool.n_free_blocks(), "Total used CPU memory:", psutil.Process(os.getpid()).memory_info().rss/1024/1024/1024, "GB")
# output: 0, 8G

# transfer from pinned memory to device memory, this should not need more pinned_memory!
b = cupy.asarray(a)
print("b = cupy.asarray(a)")
print("Pinned-memory Pool Free blocks:", pool.n_free_blocks(), "Total used CPU memory:", psutil.Process(os.getpid()).memory_info().rss/1024/1024/1024, "GB")
# output: 0, 16G

del a, b
print("del a,b")
print("Pinned-memory Pool Free blocks:", pool.n_free_blocks(), "Total used CPU memory:", psutil.Process(os.getpid()).memory_info().rss/1024/1024/1024, "GB")
# output: 1, 16G

# the extra pinned_memory resist even if all variables are freeed
pool.free_all_blocks()
print("free_all_blocks()")
print("Pinned-memory Pool Free blocks:", pool.n_free_blocks(), "Total used CPU memory:", psutil.Process(os.getpid()).memory_info().rss/1024/1024/1024, "GB")
# output: 0, 8G
```

### Installation

Conda-Forge (`conda install ...`)

### Environment

```
OS                           : Linux-5.4.143.bsk.7-amd64-x86_64-with-glibc2.31
Python Version               : 3.10.9
CuPy Version                 : 12.3.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.26.3
SciPy Version                : 1.11.4
Cython Build Version         : 0.29.36
Cython Runtime Version       : None
CUDA Root                    : /usr/local/cuda
nvcc PATH                    : /usr/local/cuda/bin/nvcc
CUDA Build Version           : 11080
CUDA Driver Version          : 11080
CUDA Runtime Version         : 11080
cuBLAS Version               : (available)
cuFFT Version                : 10900
cuRAND Version               : 10300
cuSOLVER Version             : (11, 4, 1)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 8)
Thrust Version               : 101501
CUB Build Version            : 101501
Jitify Build Version         : 6fc36c7
cuDNN Build Version          : 8800
cuDNN Version                : 8600
NCCL Build Version           : 21904
NCCL Runtime Version         : 21501
cuTENSOR Version             : 10700
cuSPARSELt Build Version     : None
Device 0 Name                : NVIDIA A100-SXM4-80GB
Device 0 Compute Capability  : 80
Device 0 PCI Bus ID          : 0000:8E:00.0
```

**Comment by leofang** [comment]
This should be fixed in v13rc1, could you give it a try?
https://github.com/cupy/cupy/releases/tag/v13.0.0rc1
https://docs.cupy.dev/en/latest/upgrade.html#change-in-cupy-array-cupy-asarray-cupy-asanyarray-behavior

**Comment by murdock-grewar** [comment]
See #8111 . Pinned memory is not released by garbage collection.

**Comment by leofang** [comment]
> See #8111 . Pinned memory is not released by garbage collection.

Answered there. Also, that issue is independent of this one. This one is on "extra" pinned memory being used, which was the behavior before the v13rc1 fix landed. (Before the fix, we unconditionally copied host data into internal buffer, regardless whether the data resides in pageable or page-locked/pinned memory.)

**Comment by leofang** [comment]
@jemiryguo does cupy v13.0 address this issue for you?

**Comment by jemiryguo** [comment]
> @jemiryguo does cupy v13.0 address this issue for you?

Yes, this is already fixed in v13.0.0.

**Comment by murdock-grewar** [comment]
Can this be backported to cupy v12?
